### PR TITLE
CAT-2157 Extension for Distance to

### PR DESCRIPTION
**Pull request by robertriedl**

### CODE REVIEW

File: catroid/src/main/java/org/catrobat/catroid/formulaeditor/InternFormulaParser.java [hunk 1]
```diff
@@ -298,6 +306,49 @@ private FormulaElement collision() throws InternFormulaParserException {
...
+
+		for (Sprite sprite : ProjectManager.getInstance().getSceneToPlay().getSpriteList()) {
+			if (sprite.getName().compareTo(currentSpriteName) == 0) {
+				spriteCount++;
+			}
+		}
+
+		if (!(spriteCount == 1)) {
+			throw new InternFormulaParserException("Parse Error, Sprite was not found");
+		}
```
Comment (robertpainsi) [inline catroid/src/main/java/org/catrobat/catroid/formulaeditor/InternFormulaParser.java]: `Scene.containsSpriteBySpriteName` or `Scene.getSpriteBySpriteName` (Those method names though… :worried:)
Use these methods everywhere possible in this PR.

File: catroid/src/main/java/org/catrobat/catroid/formulaeditor/InternToExternGenerator.java [hunk 2]
```diff
@@ -300,6 +298,11 @@ private String generateExternStringFromToken(InternToken internToken, boolean tr
 				String collisionTag = CatroidApplication.getAppContext().getString(R.string
 						.formula_editor_function_collision);
 				return collisionTag + "(" + internToken.getTokenStringValue() + ")";
+			case DISTANCE_TO_TOUCH_POSITION:
+			case DISTANCE_TO_SPRITE_POSITION:
+				String distanceTo = CatroidApplication.getAppContext().getString(R.string
+						.formula_editor_object_distance_to);
+				return distanceTo + "(" + internToken.getTokenStringValue() + ")";
```
Comment (robertpainsi) [inline catroid/src/main/java/org/catrobat/catroid/formulaeditor/InternToExternGenerator.java]: RTL languages

File: catroid/src/main/java/org/catrobat/catroid/formulaeditor/InternToken.java [hunk 3]
```diff
@@ -78,6 +78,18 @@ public void updateCollisionFormulaToVersion(float catroidLanguageVersion) {
...
+			tokenStringValue = newName;
+		}
+	}
+
+	public void updateDistanceToSpritePositionFormula(String oldName, String newName) {
+		if (internTokenType == InternTokenType.DISTANCE_TO_SPRITE_POSITION && tokenStringValue.equals(oldName)) {
+			tokenStringValue = newName;
+		}
+	}
+
```
Comment (robertpainsi) [inline catroid/src/main/java/org/catrobat/catroid/formulaeditor/InternToken.java]: Used?

File: catroid/src/main/java/org/catrobat/catroid/content/Look.java [hunk 4]
```diff
@@ -319,14 +319,20 @@ public void setYInUserInterfaceDimensionUnit(float y) {
 		setY(y - getHeight() / 2f);
 	}
 
-	public float getDistanceToTouchPositionInUserInterfaceDimensions() {
+	public double getDistanceToTouchPositionInUserInterfaceDimensions() {
```
Comment (robertriedl) [inline catroid/src/main/java/org/catrobat/catroid/content/Look.java]: `TouchUtil` is needed to get the last touch position

File: catroid/src/main/java/org/catrobat/catroid/formulaeditor/InternFormulaParser.java (hunk 1 above)
Comment (robertriedl) [inline catroid/src/main/java/org/catrobat/catroid/formulaeditor/InternFormulaParser.java]: These methods in `Scene` stop, when the first sprite is found, but in this case the currentSprite and the secondSprite have to be found

Comment (robertpainsi) [inline catroid/src/main/java/org/catrobat/catroid/formulaeditor/InternFormulaParser.java]: Always keep [premature optimization](https://en.wikipedia.org/wiki/Program_optimization#When_to_optimize) in mind.

However, you hinted me to something else.
IMO it doesn't make sense to search for the `CurrentSprite` in the `SceneToPlay`* anyway. A lot of code builds on the premise that the opened Scene,Sprite in the UI is the `getCurrentScene`,`getCurrentSprite`.
Searching for the second sprite however makes sense. If you've deleted the second sprite, the formula becomes invalid.**
What do you think?

\* `SceneToPlay == CurrentScene`. I'll create a PR for that.
\** I'm not 100% sure if *a parser* should check this, but it's not important for this PR.

File: catroid/src/main/java/org/catrobat/catroid/formulaeditor/FormulaElement.java [hunk 5]
```diff
@@ -270,10 +277,34 @@ public Object interpretRecursive(Sprite sprite) {
...
 
+	private Object interpretDistanceToSpritePosition(Sprite currentSprite, String value) {
+		String secondSpriteName = value;
+		Sprite secondSprite;
+
+		try {
+			secondSprite = ProjectManager.getInstance().getSceneToPlay().getSpriteBySpriteName(secondSpriteName);
+		} catch (Resources.NotFoundException exception) {
+			return 0d;
+		}
```
Comment (robertpainsi) [inline catroid/src/main/java/org/catrobat/catroid/formulaeditor/FormulaElement.java]: I should've tested it in Scratch and also do more research… returning `0` is fine (I guess). Many calculations in the FormulaElement return `0` if there is some kind of undefined intermediate result or state. The `InterpretationException` is just used if the **result** of the Formula is NaN.

File: catroid/src/main/java/org/catrobat/catroid/formulaeditor/FormulaElement.java [hunk 6]
```diff
@@ -270,10 +277,43 @@ public Object interpretRecursive(Sprite sprite) {
...
+	}
+
+	private Object interpretDistanceToSpritePosition(Sprite currentSprite, String secondSpriteName) {
+		Sprite secondSprite;
+
+		try {
+			secondSprite = ProjectManager.getInstance().getSceneToPlay().getSpriteBySpriteName(secondSpriteName);
+		} catch (Resources.NotFoundException exception) {
+			return 0d;
+		}
```
Comment (robertpainsi) [inline catroid/src/main/java/org/catrobat/catroid/formulaeditor/FormulaElement.java]: None of the methods in the try block throw a `NotFoundException`.

File: catroid/src/main/java/org/catrobat/catroid/ui/fragment/FormulaEditorCategoryListFragment.java [hunk 7]
```diff
@@ -76,16 +76,15 @@
 	private CategoryListAdapter adapter;
 
 	private static final int[] OBJECT_GENERAL_PROPERTIES_ITEMS = { R.string.formula_editor_object_transparency,
-			R.string.formula_editor_object_brightness, R.string.formula_editor_object_color/*,
-			R.string.formula_editor_object_distance_to*/ };
```
Comment (robertpainsi) [inline catroid/src/main/java/org/catrobat/catroid/ui/fragment/FormulaEditorCategoryListFragment.java]: Just to be sure, the `distance to` feature hasn't been enabled in any release yet, right?
So we don't have to cover backward compatibility.

**Comment by robertpainsi** [comment]
The value `touch position` will be (unnecessarily) stored in the XML.
The value stored in the XML will be shown to the user, independently of the users language.

File: catroid/src/main/java/org/catrobat/catroid/ui/fragment/FormulaEditorCategoryListFragment.java (hunk 7 above)
Comment (robertriedl) [inline catroid/src/main/java/org/catrobat/catroid/ui/fragment/FormulaEditorCategoryListFragment.java]: no, distance to hasn't been in any release

**Pull request closed without merging**